google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -321,50 +357,475 @@ private boolean isAlive() {
 
   @Override
   public void sendAckOperations(
-      List<String> acksToSend, List<PendingModifyAckDeadline> ackDeadlineExtensions) {
-    ApiFutureCallback<Empty> loggingCallback =
-        new ApiFutureCallback<Empty>() {
-          @Override
-          public void onSuccess(Empty empty) {
-            ackOperationsWaiter.incrementPendingCount(-1);
-          }
+      List<ModackWithMessageFuture> modackWithMessageFutures,
+      List<AckIdMessageFuture> ackIdMessageFutures) {
+    // Process modacks first
+    Set<String> failedAckIds = sendModacks(modackWithMessageFutures);
+
+    // Remove failed modacks from acks
+    ackIdMessageFutures.removeIf(
+        ackIdMessageFuture -> {
+          return failedAckIds.contains(ackIdMessageFuture.getAckId());
+        });
+
+    sendAcks(ackIdMessageFutures);
+  }
 
-          @Override
-          public void onFailure(Throwable t) {
-            ackOperationsWaiter.incrementPendingCount(-1);
-            Level level = isAlive() ? Level.WARNING : Level.FINER;
-            logger.log(level, "failed to send operations", t);
+  private class RetryModacksFailedModacks {
+    private List<ModackWithMessageFuture> retryModackWithMessageFutures;
+    private Set<String> failedAckIds;
+
+    RetryModacksFailedModacks(
+        List<ModackWithMessageFuture> retryModackWithMessageFutures, Set<String> failedAckIds) {
+      this.retryModackWithMessageFutures = retryModackWithMessageFutures;
+      this.failedAckIds = failedAckIds;
+    }
+
+    public List<ModackWithMessageFuture> getRetryModackWithMessageFutures() {
+      return retryModackWithMessageFutures;
+    }
+
+    public Set<String> getFailedAckIds() {
+      return failedAckIds;
+    }
+  }
+
+  private Map<String, String> getMetadataMapFromThrowable(Throwable t) {
+    // This converts a Throwable (from a "OK" grpc response) to a map of metadata
+    // will be of the format:
+    // {
+    //    "ACK-ID-1": "PERMANENT_*",
+    //    "ACK-ID-2": "TRANSIENT_*"
+    // }
+    com.google.rpc.Status status = StatusProto.fromThrowable(t);
+    Map<String, String> metadataMap = new HashMap<>();
+    if (status != null) {
+      for (Any any : status.getDetailsList()) {
+        if (any.is(ErrorInfo.class)) {
+          try {
+            ErrorInfo errorInfo = any.unpack(ErrorInfo.class);
+            metadataMap = errorInfo.getMetadataMap();
+          } catch (Throwable throwable) {
```

Comment:
Should we:

- Log something here?
- Capture the status and propagate to the client as one of the ENUM values on top of the PR?

Reply:
Would we want to fail all AckIds in the request that lead to this throwable?

Reply:
Discussed offline yesterday. Answer is yes.